### PR TITLE
C/C++: Detect ambiguous assignment of comparison results

### LINKED ISSUE #22286: C: Implicit compare-then-assign in branch conditions should be flagged
Implicit compare-then-assign in branch conditions is buggy since developers often mean assign-then-compare, but sometimes actually mean compare-then-assign. GCC's -Wparentheses was originally meant to catch assignment in place of comparison, requiring an extra set of parentheses to turn this off. This had the happy coincidence of making developers explicit about assign-then-compare vs compare-then-assign.

An outer level of extra parentheses will inhibit -Wparentheses warnings. This often results in assign-then-compare being made explicit, but instead of turning `if (x = foo() < 0)` into `if ((x = foo()) < 0)`, a developer might write `if ((x = foo() < 0))`, which turns off the warning, without fixing the problem. This happened in openzfs/zfs#18874. There are other potential variations, such as `if ((x = (foo()) < 0))`, which also suppresses GCC's warning, but fails to actually do anything since the intended explicit parentheses to specify compare-then-assign are around the right operand of the boolean operator, rather than around the boolean operator, yet we have the additional parentheses needed to silence GCC's -Wparentheses. In the `if ((x = (foo()) < 0))` case, the intent was to make compare-then-assign explicit, and a typo caused it to fail to become explicit. That is not a bug, but it makes it unclear what the developer intended, which is problematic in itself.

A CodeQL check to detect this condition will likely be merged into openzfs/zfs soon:

https://github.com/openzfs/zfs/pull/18899

It has also been subjected to a test suite of 19 true positives and 17 true negatives to verify its behavior. It successfully detected all true positives and fails to detect any true negatives. It has also been applied not only to the OpenZFS codebase, but also the Linux kernel and curl codebases, where it had zero detections. Related queries in CodeQL were also run against the test suite, but had zero detections. In specific, all of these queries failed to detect a single test case in the test suite:

* cpp/assign-where-compare-meant
* cpp/compare-where-assign-meant
* cpp/comparison-precedence
* cpp/operator-precedence-logic-error-when-use-bool-type
* cpp/operator-precedence-logic-error-when-use-bitwise-logical-operations

It is unclear whether I should include the test suite in my OpenZFS PR, but I would be happy to include it in a CodeQL PR if upstream is willing to take it.

For reference, the test suite is simply this:

```
/* Synthetic fixtures for ambiguous assignment-of-comparison query.
 * TP = true positive (must flag). TN = true negative (must not flag).
 */

int foo(void);
int bar(void);
void *getp(void);

/* ========== TRUE POSITIVES (ambiguous / GCC-silent bugs) ========== */

/* TP1: classic PR-style bug — outer parens silence GCC */
int tp1(void) {
    int x;
    if ((x = foo() < 0))
        return x;
    return 0;
}

/* TP2: while form */
int tp2(void) {
    int x;
    while ((x = foo() != 0))
        return x;
    return 0;
}

/* TP3: no outer parens — GCC also warns; still the same pattern */
int tp3(void) {
    int x;
    if (x = foo() < 0)
        return x;
    return 0;
}

/* TP4: for condition */
int tp4(void) {
    int x;
    for (; (x = bar() <= 0); )
        return x;
    return 0;
}

/* TP5: && chain — assignment-of-comparison is an operand */
int tp5(void) {
    int x, y = 1;
    if ((x = foo() < 0) && y)
        return x;
    return 0;
}

/* TP6: ternary */
int tp6(void) {
    int x;
    return (x = foo() > 0) ? x : 0;
}

/* TP7: pointer equality on RHS, not parenthesized as comparison */
int tp7(void) {
    int x;
    if ((x = getp() == 0))
        return x;
    return 0;
}

/* TP8: (foo()) only parenthesizes the call, not the comparison — still ambiguous */
int tp8(void) {
    int x;
    if ((x = (foo()) < 0))
        return x;
    return 0;
}

/* TP9: do-while */
int tp9(void) {
    int x;
    do {
        x = 1;
    } while ((x = foo() < 0));
    return x;
}

/* TP10: || chain */
int tp10(void) {
    int x, y = 0;
    if (y || (x = foo() > 0))
        return x;
    return 0;
}

/* TP11: compound += with unparenthesized comparison */
int tp11(void) {
    int x = 0;
    if ((x += foo() < 0))
        return x;
    return 0;
}

/* TP12: compound -= */
int tp12(void) {
    int x = 0;
    while ((x -= bar() != 0))
        return x;
    return 0;
}

/* TP13: compound |= */
int tp13(void) {
    int x = 0;
    if ((x |= foo() > 0))
        return x;
    return 0;
}

/* TP14: compound &= in && chain */
int tp14(void) {
    int x = ~0, y = 1;
    if ((x &= foo() < 0) && y)
        return x;
    return 0;
}

/* TP15: assignment of unparenthesized comparison nested under outer comparison
 * (still ambiguous what was assigned; simplified rule flags this) */
int tp15(void) {
    int x;
    if ((x = foo() < 0) == 1)
        return x;
    return 0;
}

/* TP16: compound *= */
int tp16(void) {
    int x = 1;
    if ((x *= foo() != 0))
        return x;
    return 0;
}

/* TP17: compound ^= in ternary */
int tp17(void) {
    int x = 0;
    return (x ^= foo() < 0) ? x : 0;
}

/* TP18: compound <<= */
int tp18(void) {
    int x = 1;
    if ((x <<= bar() > 0))
        return x;
    return 0;
}

/* ========== TRUE NEGATIVES (explicit decision) ========== */

/* TN1: explicit assign-then-compare */
int tn1(void) {
    int x;
    if ((x = foo()) < 0)
        return x;
    return 0;
}

/* TN2: explicit assign then != */
int tn2(void) {
    int x;
    while ((x = foo()) != 0)
        return x;
    return 0;
}

/* TN3: intentional assign of non-comparison, parenthesized */
int tn3(void) {
    int x;
    if ((x = foo()))
        return x;
    return 0;
}

/* TN4: explicit compare-then-assign — parens around the comparison */
int tn4_explicit_bool_assign(void) {
    int flag;
    if ((flag = (foo() < 0)))
        return flag;
    return 0;
}

/* TN5: extra redundant parens around assign-then-compare — still explicit */
int tn5(void) {
    int x;
    if (((x = foo()) < 0))
        return x;
    return 0;
}

/* TN6: assign constant used as condition — different pattern, not ours */
int tn6(void) {
    int x;
    if ((x = 0))
        return x;
    return 0;
}

/* TN7: plain comparison */
int tn7(void) {
    int x = foo();
    if (x < 0)
        return x;
    return 0;
}

/* TN8: explicit compare-then-assign without outer double-paren style */
int tn8(void) {
    int x;
    if (x = (foo() < 0))  /* GCC may warn about assign as truth; still explicit intent */
        return x;
    return 0;
}

/* TN9: assign then compare with >= */
int tn9(void) {
    int x;
    for (; (x = bar()) >= 0; )
        return x;
    return 0;
}

/* TN10: logical and of two explicit assign-then-compares */
int tn10(void) {
    int x, y;
    if ((x = foo()) < 0 && (y = bar()) < 0)
        return x + y;
    return 0;
}

/* TN11: compound += then compare (explicit assign-then-compare) */
int tn11(void) {
    int x = 0;
    if ((x += foo()) < 0)
        return x;
    return 0;
}

/* TN12: compound += of parenthesized comparison (explicit compare-then-assign) */
int tn12(void) {
    int x = 0;
    if ((x += (foo() < 0)))
        return x;
    return 0;
}

/* TN13: compound &= of non-comparison */
int tn13(void) {
    int x = ~0;
    if ((x &= foo()))
        return x;
    return 0;
}

/* TN14: compound |= then compare */
int tn14(void) {
    int x = 0;
    while ((x |= bar()) != 0)
        return x;
    return 0;
}

/* TN15: compound -= of parenthesized comparison */
int tn15(void) {
    int x = 0;
    if ((x -= (foo() != 0)))
        return x;
    return 0;
}

/* TN16: compound <<= then compare */
int tn16(void) {
    int x = 1;
    if ((x <<= bar()) > 0)
        return x;
    return 0;
}

/* TP19: bug pattern inside a macro expansion — must still be reported */
#define BAD_CHECK(x) if (((x) = foo() < 0)) return (x)
int tp19_macro(void) {
    int x;
    BAD_CHECK(x);
    return 0;
}

/* TN17: explicit assign-then-compare inside a macro — must not report */
#define GOOD_CHECK(x) if (((x) = foo()) < 0) return (x)
int tn17_macro(void) {
    int x;
    GOOD_CHECK(x);
    return 0;
}
```

Is this something that should go into a PR, or would upstream prefer to handle this its own way?

### PR DESCRIPTION
**Pull request by theinfosecguy**
Adds `cpp/ambiguous-assignment-of-comparison` to flag ambiguous assignments when the assignment result is used as a truth value, such as:

```c
if ((status = read_status() < 0))
```

The query distinguishes this from explicitly grouped assign-then-compare and compare-then-assign expressions. It includes C and C++ tests, query help, and query-suite integration.

Local targeted and neighboring tests pass. The motivating regression is detected, and a run against `git/git` produced no alerts.

Fixes #22286

### CODE REVIEW

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+int read_status();
+
+int check_status() {
+  int status;
+  if ((status = read_status() < 0)) // BAD: assigns the comparison result.
+    return status;
+
+  if ((status = read_status()) < 0) // GOOD: assigns first, then compares.
```
Comment (ryao) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp:8]: This isn't the only good variant. The following is good too, because it shows that the developer really meant it.

```
  if ((status = (read_status() < 0))) // GOOD: explicitly assigns the comparison result.
    return status;
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  exists(Cast cast | cast = comparison.getConversion+() and not cast.isImplicit())
+}
+
+from Assignment assignment, ComparisonOperation comparison
+where
+  assignment.getRValue() = comparison and
+  not isExplicitlyGrouped(comparison) and
+  occursInCondition(assignment) and
+  // Assigning a comparison result to a Boolean is normally intentional.
+  not assignment.getLValue().getUnspecifiedType() instanceof BoolType and
```
Comment (ryao) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:66]: While this is normally intentional, I believe `not isExplicitlyGrouped(comparison)` precludes this.

**Review by ryao — COMMENTED** [review]
While I am happy you did the work for me to get this into a PR and made what appear to be improvements, would you add an `Original-patch-by:` to the commit message to credit my prior work?

Also, have you run your variant against any major corpora (e.g. Linux, curl, OpenZFS) to verify the lack of FPs in production code, like I did with the original version? If it helps:

https://docs.github.com/en/code-security/how-tos/find-and-fix-code-vulnerabilities/scan-from-the-command-line/download-databases

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp [hunk 3]
```diff
@@ -0,0 +1,12 @@
+int read_status();
+
+int check_status() {
+  int status;
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp]: Lets have a version of the test without the double-brackets as well.
```suggestion
  int status;

  if (status = read_status() < 0) // BAD: assigns the comparison result.
    return status;
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp]: I agree, and it's good to have test casess showing what we do in common cases like this.
```suggestion

  if (status = (read_status()) < 0) // GOOD: the extra brackets suggests this is intended.
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  if ((status = read_status() < 0)) // BAD: assigns the comparison result.
+    return status;
+
+  if ((status = read_status()) < 0) // GOOD: assigns first, then compares.
+    return status;
+
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp]:
```suggestion

  bool  result;

  if (result = read_status() < 0) // GOOD: the type of result suggests this is intended.
 
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp:8]: That's true, but we don't need to show every possibility in the `.qhelp` example.  The test is the place to be thorough.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp [hunk 5]
```diff
@@ -0,0 +1,12 @@
+int read_status();
+
+int check_status() {
+  int status;
+  if ((status = read_status() < 0)) // BAD: assigns the comparison result.
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp]: I don't think the double brackets are communicating anything to the reader here.
```suggestion
  if (status = read_status() < 0) // BAD: assigns the comparison result.
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+/** Gets a condition that controls branching. */
+private Expr getACondition() {
+  result = any(IfStmt s).getCondition()
+  or
+  result = any(WhileStmt s).getCondition()
+  or
+  result = any(DoStmt s).getCondition()
+  or
+  result = any(ForStmt s).getCondition()
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]:
```suggestion
  result = any(Loop s).getCondition()
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  or
+  result = any(ConditionalExpr e).getCondition()
+}
+
+/**
+ * Holds if `assignment` occurs within a condition that controls branching.
+ *
+ * This includes nested expressions, such as function arguments and either operand of a comma
+ * expression, because the ambiguous syntax still occurs within the condition.
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: What's the motivating case here?  I think there's a good argument we want this for things like the right argument of a comma expression, but I'm not convinced we care about the arguments to function calls that just happens to be located somewhere inside an `if` expression.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 6 above)
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: We can simplify this a little:
```suggestion
  result = any(Loop s).getCondition()
```

**Review by geoffw0 — COMMENTED** [review]
I've reviewed the code and docs from a technical perspective (I haven't looked through all the test cases yet).  This looks quite promising.  I've also done a mass (MRVA) run and found quite a high rate of true positive results!

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:66]: Why is that? It seems like the types of the expressions and the bracketing are mostly independent concerns.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  result = any(IfStmt s).getCondition()
+  or
+  result = any(WhileStmt s).getCondition()
+  or
+  result = any(DoStmt s).getCondition()
+  or
+  result = any(ForStmt s).getCondition()
+  or
+  result = any(ConditionalExpr e).getCondition()
+}
```
Comment (geoffw0) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:25]: I have an AI suggestion, which is to relax this function from expressions that "control branching" to expressions "whose value is used as a truth value", adding these cases:
```suggestion
  or
  result = any(UnaryLogicalOperation op).getAnOperand()
  or
  result = any(BinaryLogicalOperation op).getAnOperand()
}
```

Looking at your code again, the absence of these two cases might be the main reason you found yourself needing `occursInCondition`.  With them, we might be better off without `occursInCondition` at all???

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp (hunk 5 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.cpp]: Addressed.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 6 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: Addressed.

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+int logical_or() {
+  int value;
+  if (get_other_value() || (value = get_value() > 0)) // $ Alert // BAD
+    return value;
+  return 0;
+}
+
+int nested_comparison() {
+  int value;
+  if ((value = get_value() < 0) == 1) // $ Alert // BAD
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp]: This would be an odd thing for someone to write, the `== 1` suggests the developer is expecting the numeric result of `get_value()` to be compared with `1`, but neither bracketing would actually accomplish that.  Perhaps a more natural thing they might write would be:
```suggestion
  if ((value = get_value() < 0) == false) // $ Alert // BAD
```

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+int assignment_as_call_argument() {
+  int value;
+  if (identity(value = get_value() < 0)) // $ Alert // BAD: The ambiguous syntax is still in a condition.
+    return value;
+  return 0;
+}
+
+int discarded_assignment_in_comma_expression() {
+  int value;
+  if ((value = get_value() < 0, get_other_value())) // $ Alert // BAD: Still ambiguous syntax in a condition.
```
Comment (geoffw0) [inline cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp]: This is another slightly odd one for someone to write.  If the developer knows the left side of the comma expression is discarded, why did they put `< 0` there?  Most likely because they *do* know they're assigning the comparison result to `value` I'd guess, so it's maybe not ideal to flag this.

Unless maybe the `,` itself is unintended by the developer?

I think the case we're really interested in flagging is the RHS of the comma expression:

```suggestion
  if ((value = get_value() < 0, get_other_value())) // $ SPURIOUS: Alert // GOOD: Seems like assigning the comparison result would have to be on purpose in this case?

  if ((get_other_value(), value = get_value() < 0)) // $ Alert // BAD: Still ambiguous syntax in a condition.
```

**Review by geoffw0 — COMMENTED** [review]
Nice thorough tests.  I've nit-picked a couple of cases but they're not important ones, and you might have more insight into them than I have, from knowing the cases that motivated this query.

---

I've gained a lot of confidence in this query from reviewing it - just a few small things to discuss and decide if we want to make any changes or not.  :+1:

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp (hunk 9 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp]: Addressed. :+1:

File: cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp (hunk 10 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison/test.cpp]: Addressed. :+1:

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 8 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:25]: Addressed. :+1:

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 7 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: No longer relevant.

**Comment by geoffw0** [comment]
Thank you for making the changes.  The CodeQL looks good to me, as do the tests and MRVA (bulk testing) results.  I think I said it before, but I'm seeing lots of true positive results in real world data.  :)

I've requested a quick review of the `.qhelp` from the docs team, we should be able to merge soon after that.

**Comment by jclement136** [comment]
👋 Hello from docs! Just confirming that I've added this to our review board for you to make sure it gets attention and a review soon. If you still need any assistance just reach out on Slack via the `#docs-content` channel.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 2 above)
Comment (ryao) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:66]: My reasoning was that if it is a comparison operation, then it is implicitly bool. Maybe with C++ operator overloading, this might not be the case. I had only been imagining C when I made this remark.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 11]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @name Ambiguous assignment of comparison used as truth value
```
Comment (mchammer01) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:2]: The query name is grammatically unclear, suggesting the following:
```suggestion
 * @name Ambiguous assignment of comparison result used as truth value
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ * @description Assigning the result of an unparenthesized comparison when the assignment is used
+ *              as a truth value may indicate that the assignment and comparison are grouped
+ *              incorrectly.
```
Comment (mchammer01) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql:5]: The description is complex and doesn’t closely follow the preferred “Syntax X causes behavior Y” pattern.
Suggesting something like: (let me know what you think!)

```suggestion
 * @description  Assigning the result of an unparenthesized comparison when the assignment is used
 *               as a truth value can obscure the intended grouping of the operations.
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+<overview>
+<p>Assignment operators have lower precedence than comparison operators. For example,
+<code>status = read_status() &lt; 0</code> assigns the comparison result (zero or one) to
+<code>status</code>. This can be unintended when the programmer meant to assign the return value
+first and then compare it with zero.</p>
+</overview>
+
+<recommendation>
+<p>Use parentheses to make the intended operation order explicit. To assign first and compare the
```
Comment (mchammer01) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp:14]: Nit: use more natural language:
```suggestion
<p>Use parentheses to make the intended order of operations explicit. To assign first and compare the
```

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+assigned value, parenthesize the assignment. To intentionally assign the comparison result,
+parenthesize the comparison. An explicit cast around the comparison also makes that order clear.</p>
+</recommendation>
+
+<example>
+<p>In the first condition, <code>status</code> receives either zero or one instead of the value
+returned by <code>read_status</code>. The second condition explicitly performs the assignment
```
Comment (mchammer01) [inline cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp:21]: Change  `read_status`  to  `read_status()`  when referring to the function call
```suggestion
returned by <code>read_status()</code>. The second condition explicitly performs the assignment
```

File: cpp/ql/src/change-notes/2026-08-13-ambiguous-assignment-of-comparison.md [hunk 15]
```diff
@@ -0,0 +1,5 @@
+---
+category: newQuery
+---
+* Added a new query, `cpp/ambiguous-assignment-of-comparison`, to detect assignments of
+  unparenthesized comparison results when the assignment is used as a truth value.
```
Comment (mchammer01) [inline cpp/ql/src/change-notes/2026-08-13-ambiguous-assignment-of-comparison.md:5]: Nit: Whilst this is technically accurate, the wording is dense. “Assignments of unparenthesized comparison results” takes a moment to parse. What about:
```suggestion
* Added a new query, `cpp/ambiguous-assignment-of-comparison`, to detect potentially
  ambiguous expressions where a comparison result is assigned to a variable and the
  assignment is used as a truth value.
```

**Review by mchammer01 — DISMISSED** [review]
@theinfosecguy 👋🏻 - Hi from Technical Content (used to be Docs).  Overall, this PR looks good, so I’m approving it ✨ 
I’ve left a few suggestions to make the query help, metadata, and change note easier to read. It would be great if some of these could be applied or tweaked before merging, but I don’t think they need to block the PR. Thanks for adding clear examples and references 🙏

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 12 above)
Comment (geoffw0) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: I'll offer an alternative:
```suggestion
 * @description Using an assignment of an unparenthesized comparison as
 *              a truth value may indicate unintended operator grouping.
```
But I'm happy with either suggestion, or something similar.

**Comment by geoffw0** [comment]
@theinfosecguy I'm ready to merge this once you've responded to @mchammer01 's review - e.g. by accepting the suggestions.  Nothing else  needs doing now.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 11 above)
Comment (theinfosecguy) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: Applied the suggested query name, thanks.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql (hunk 12 above)
Comment (theinfosecguy) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.ql]: Applied Geoff's alternative wording, thanks.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp (hunk 13 above)
Comment (theinfosecguy) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp]: Applied the suggested wording, thanks.

File: cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp (hunk 14 above)
Comment (theinfosecguy) [reply cpp/ql/src/Likely Bugs/Likely Typos/AmbiguousAssignmentOfComparison.qhelp]: Updated the function call, thanks.

File: cpp/ql/src/change-notes/2026-08-13-ambiguous-assignment-of-comparison.md (hunk 15 above)
Comment (theinfosecguy) [reply cpp/ql/src/change-notes/2026-08-13-ambiguous-assignment-of-comparison.md]: Applied the suggested wording, thanks.

**Comment by theinfosecguy** [comment]
Thanks, I've applied the review suggestions and responded to each thread.

**Review by geoffw0 — APPROVED** [review]
Everything LGTM now, thank you very much @theinfosecguy for your contribution (and for addressing the pull request comments).  🎉 ✨ 🚀
